### PR TITLE
feature/voxel-world

### PR DESCRIPTION
**Pull request by Genora51**
- Interfaces to provide terrain through voxels
- Conversion from voxel field to mesh via marching cubes with a GPU compute shader
- Support for different shading based on voxel type
- Space-efficient serialization and deserialization of voxel field modifications using an octree data structure
- Chunk-based voxel loading and meshing system
- Multithreaded chunk loading

### CODE REVIEW

File: Assets/Scripts/Terrain/Meshing/MeshManager.cs [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        {
+            if (!_meshHolders.TryGetValue(chunkLocation, out var chunkHolder))
+            {
+                // No valid loaded chunk
+                return;
+            }
+            _meshHolders.Remove(chunkLocation);
+            // Recycle the chunk
+            // TODO: Provide method to cleanup recycled chunks to save memory
+            chunkHolder.MeshFilter.mesh = null;
```
Comment (SlyCedix) [inline Assets/Scripts/Terrain/Meshing/MeshManager.cs:180]: Doesn't properly dispose of mesh, causing a memory leak

Fixed in [188b524](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/commit/873680db00ea6efae350a36c00a9c0313eba5fd4)

File: Assets/Scripts/Terrain/Chunk/ChunkManager.cs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+        /// <summary>
+        /// Decide which chunks to load and recycle chunks no longer needed.
+        /// </summary>
+        private void QueueNewChunkLoads()
+        {
+            var chunksToLoad = new HashSet<Vector3Int>();
+            var currentChunk = Vector3Int.FloorToInt(_chunkLoadingTarget.position / _chunkSize);
+            // Get the set of chunks we want loaded this frame
+            for (var x = -_chunkLoadDistance; x <= _chunkLoadDistance; x++)
```
Comment (SlyCedix) [inline Assets/Scripts/Terrain/Chunk/ChunkManager.cs:151]: Regenerates list of chunk offsets every frame, unnecessarily.

Fixed in [873680d](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/commit/873680db00ea6efae350a36c00a9c0313eba5fd4)

File: Assets/Scripts/Terrain/Meshing/MeshManager.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        /// <summary>
+        /// Handle the ChunkLoaded event.
+        /// </summary>
+        /// <param name="chunkLocation">The location of the loaded chunk.</param>
+        private void OnChunkLoaded(Vector3Int chunkLocation)
+        {
+            // Mesh chunk by location
+            for (var i = 0; i < 8; i++)
```
Comment (SlyCedix) [inline Assets/Scripts/Terrain/Meshing/MeshManager.cs:89]: Remeshes chunks that have already been meshed, requiring 3.6 times more meshes to be processed than necessary

Fixed naively in [0dd6a20](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/commit/0dd6a2001cdc36736f3207f59bed8f4777d0753f)

Can be improved further by ensuring that `OnChunkLoaded` is only called when the edge chunk data is present in `LoadedChunks` by implementing a system for chunk preloading

File: Assets/Scripts/Terrain/Meshing/MeshManager.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        }
+        
+        public void Awake()
+        {
+            _chunkManager = GetComponent<ChunkManager>();
+        }
+
+        public void Update()
+        {
+            // HACK: Mesh at most one chunk loaded this frame
```
Comment (SlyCedix) [inline Assets/Scripts/Terrain/Meshing/MeshManager.cs:51]: Only meshes one chunk per frame, far too slow for smooth gameplay at large render distances

Fixed in [188b524](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/commit/188b524ad54bbb5dd7df10baf0fed3ae425d7efc)

File: Assets/Scripts/Terrain/Chunk/ChunkManager.cs (hunk 2 above)
Comment (SlyCedix) [inline Assets/Scripts/Terrain/Chunk/ChunkManager.cs:151]: Order chunks are loaded in are not based on proximity to ChunkLoadTarget, causing chunks near the player to remain unloaded for a long time.

Fixed in [188b524](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/commit/188b524ad54bbb5dd7df10baf0fed3ae425d7efc)

**Review by SlyCedix — APPROVED** [review]
Only glaring issue is the memory leak caused by meshes not being disposed of, this is fixed in [feature/voxel-world-optimizations](https://github.com/Rebirth-of-the-Night/Rebirth-of-the-Night-SA/tree/feature/voxel-world-optimizations), so approving.

**Review by yungnickyoung — APPROVED** [review]
Sly nailed most of my big concerns. I think there are still some optimizations to be made, but that stuff can be done on the optimization branch. This looks like a great starting point. Good work :)

**Pull request merged**